openedx/core/djangoapps/user_api/accounts/api.py

```diff
@@ -0,0 +1,219 @@
+            `username` is not specified)
+        AccountNotAuthorized: the requesting_user does not have access to change the account
+            associated with `username`
+        AccountUpdateError: the update could not be completed, usually due to validation errors
+            (for example, read-only fields were specified or field values are not legal)
+
+    """
+    if username is None:
+        username = requesting_user.username
+
+    existing_user, existing_user_profile = _get_user_and_profile(username)
+
+    if requesting_user.username != username:
+        raise AccountNotAuthorized()
+
+    # If user has requested to change email, we must call the multi-step process to handle this.
+    # It is not handled by the serializer (which considers email to be read-only).
+    new_email = None
+    if "email" in update:
+        new_email = update["email"]
+        del update["email"]
+
+    # If user has requested to change name, store old name because we must update associated metadata
+    # after the save process is complete.
+    old_name = None
+    if "name" in update:
+        old_name = existing_user_profile.name
+
+    # Check for fields that are not editable. Marking them read-only causes them to be ignored, but we wish to 400.
+    read_only_fields = set(update.keys()).intersection(
+        AccountUserSerializer.Meta.read_only_fields + AccountLegacyProfileSerializer.Meta.read_only_fields
+    )
+
+    # Build up all field errors, whether read-only, validation, or email errors.
+    field_errors = {}
+
+    if read_only_fields:
+        for read_only_field in read_only_fields:
+            field_errors[read_only_field] = {
+                "developer_message": "This field is not editable via this API",
+                "user_message": _("Field '{field_name}' cannot be edited.".format(field_name=read_only_field))
+            }
+            del update[read_only_field]
+
+    user_serializer = AccountUserSerializer(existing_user, data=update)
+    legacy_profile_serializer = AccountLegacyProfileSerializer(existing_user_profile, data=update)
+
+    for serializer in user_serializer, legacy_profile_serializer:
+        field_errors = _add_serializer_errors(update, serializer, field_errors)
+
+    # If the user asked to change email, validate it.
+    if new_email:
+        try:
+            validate_new_email(existing_user, new_email)
+        except ValueError as err:
+            field_errors["email"] = {
+                "developer_message": "Error thrown from validate_new_email: '{}'".format(err.message),
+                "user_message": err.message
+            }
+
+    # If we have encountered any validation errors, return them to the user.
+    if field_errors:
+        raise AccountUpdateError({"field_errors": field_errors})
+
+    # If everything validated, go ahead and save the serializers.
+    for serializer in user_serializer, legacy_profile_serializer:
```

Comment:
Yes, I will wrap the remainder of the bits with an AccountUpdateError.

I wonder if I should create a specific type of error for validation errors (as opposed to passed validation, but error occurred during update). Perhaps AccountValidationError? What do you think @andy-armstrong?